client/client.go

```diff
@@ -830,22 +807,27 @@ func (client *Client) ConfigureGoogleAds(opts config.GoogleSearchAdsOptions) {
 	client.googleAdsOptionsLock.Unlock()
 }
 
-// initBalancer takes hosts from cfg.ChainedServers and it uses them to create a
-// balancer. Returns the new dialers.
-func (client *Client) initBalancer(proxies map[string]*commonconfig.ProxyConfig) ([]balancer.Dialer, error) {
+// initOrchestrator takes hosts from cfg.ChainedServers and it uses them to create a
+// new orchestrator. Returns the new dialers.
+func (client *Client) initOrchestrator(proxies map[string]*commonconfig.ProxyConfig) ([]orchestrator.Dialer, error) {
 	if len(proxies) == 0 {
-		return nil, fmt.Errorf("No chained servers configured, not initializing balancer")
+		return nil, fmt.Errorf("no chained servers configured, not initializing orchestrator")
 	}
 
 	chained.PersistSessionStates(client.configDir)
 	dialers := chained.CreateDialers(client.configDir, proxies, client.user)
-	client.bal.Reset(dialers)
-
-	go func() {
-		for hasSucceeding := range client.bal.HasSucceedingDialer {
-			client.statsTracker.SetHasSucceedingProxy(hasSucceeding)
-		}
-	}()
-
+	dialer, err := orchestrator.NewWithCallback(dialers, func(d orchestrator.Dialer) {
+		countryCode, country, city := proxyLoc(d)
+		client.statsTracker.SetActiveProxyLocation(
+			city,
+			country,
+			countryCode,
+		)
+		client.statsTracker.SetHasSucceedingProxy(true)
```

Comment:
This is technically slightly wrong I suppose in that if all of your dialers stop working at some point, the UI won't show that. I'll take a look at solutions.

Reply:
Fixed!